src/proxy/socks5.rs

```diff
@@ -0,0 +1,176 @@
+use crate::config::Config;
+use crate::proxy::outbound::OutboundConnection;
+use crate::proxy::Error;
+use anyhow::Result;
+use net::Ipv4Addr;
+use std::net;
+use byteorder::{BigEndian, ByteOrder};
+use crate::workload::WorkloadInformation;
+use crate::{identity, socket};
+use std::net::{IpAddr, SocketAddr};
+use tokio::io::AsyncReadExt;
+use tokio::net::{TcpListener, TcpStream};
+use tracing::{error, info, warn};
+use tokio::io::AsyncWriteExt;
+
+pub struct Socks5 {
+    cfg: Config,
+    cert_manager: identity::SecretManager,
+    workloads: WorkloadInformation,
+    listener: TcpListener,
+}
+
+impl Socks5 {
+    pub async fn new(
+        cfg: Config,
+        cert_manager: identity::SecretManager,
+        workloads: WorkloadInformation,
+    ) -> Result<Socks5, Error> {
+        let listener: TcpListener = TcpListener::bind("127.0.0.1:15080")
+            .await
+            .map_err(Error::Bind)?;
+        match socket::set_transparent(&listener) {
+            Err(_e) => info!("running without transparent mode"),
+            _ => info!("running with transparent mode"),
+        };
+
+        Ok(Socks5 {
+            cfg,
+            cert_manager,
+            workloads,
+            listener,
+        })
+    }
+
+    pub async fn run(self) {
+        let addr = self.listener.local_addr().unwrap();
+        info!("outbound listener established {}", addr);
+
+        loop {
+            // Asynchronously wait for an inbound socket.
+            let socket = self.listener.accept().await;
+            match socket {
+                Ok((stream, remote)) => {
+                    info!("accepted outbound connection from {}", remote);
+                    //let cfg = self.cfg.clone();
+                    let oc = OutboundConnection {
+                        cert_manager: self.cert_manager.clone(),
+                        workloads: self.workloads.clone(),
+                        cfg: self.cfg.clone(),
+                    };
+                    tokio::spawn(async move {
+
+                        match handle(oc, stream).await {
+                            Err(err) => log::error!("handshake error: {}", err),
+                            Ok(_) => {}
+                        }
+                    });
+                }
+                Err(e) => error!("Failed TCP handshake {}", e),
+            }
+        }
+    }
+}
+
+async fn handle(oc: OutboundConnection, mut stream: TcpStream) -> Result<(), anyhow::Error> {
+
+    let mut version = [0u8];
```

Comment:
this is very nit, but fwiw you can do 

```
let mut buf = [0u8; 2];
let [version, nmethods] = stream.read_exact(&mut buf).await?;
```

to avoid some of the `[0]` indexing and avoid 1 syscall